File: app/analytics/transport.py
"""Transport analytics: trips KPIs, route/vehicle/transporter performance."""
from __future__ import annotations

from datetime import datetime, timedelta

from sqlalchemy import case, func, select

from app import db
from app.models import Trip, Vehicle

COMPLETED = "COMPLETED"


def transport_summary(args: dict) -> dict:
    conds = [Trip.trip_status == COMPLETED]
    if args.get("start_date"):
        conds.append(Trip.departure_time >= args["start_date"])
    if args.get("end_date"):
        conds.append(Trip.departure_time <= args["end_date"] + " 23:59:59")

    on_time = case((Trip.actual_arrival <= Trip.planned_arrival, 1), else_=0)
    row = db.session.execute(
        select(
            func.count(Trip.id).label("total"),
            func.sum(on_time).label("on_time"),
            func.avg(Trip.actual_duration_hours - Trip.planned_duration_hours).label("avg_delay"),
            func.avg(Trip.distance_km / Trip.actual_duration_hours).label("avg_speed"),
            func.sum(Trip.distance_km).label("distance"),
            func.sum(Trip.fuel_liters).label("fuel"),
        ).where(*conds)
    ).one()
    m = row._mapping

    distance = m["distance"] or 0
    fuel = m["fuel"] or 0
    return {
        "total_trips": m["total"] or 0,
        "on_time_trips": m["on_time"] or 0,
        "on_time_rate": round((m["on_time"] or 0) / (m["total"] or 1) * 100, 1),
        "avg_delay_hours": round(m["avg_delay"], 1) if m["avg_delay"] is not None else None,
        "avg_speed_kmh": round(m["avg_speed"], 1) if m["avg_speed"] else None,
        "total_distance_km": round(distance, 0),
        "total_fuel_liters": round(fuel, 0),
        "fuel_efficiency_km_l": round(distance / fuel, 2) if fuel else None,
    }


def performance_by_dimension(dimension: str, limit: int = 10) -> list[dict]:
    """Trip performance grouped by route, vehicle or transporter."""
    joins_vehicle = dimension in ("vehicle", "transporter")
    if dimension == "route":
        group_cols = [Trip.origin, Trip.destination]
        label = lambda r: f"{r.origin} → {r.destination}"  # noqa: E731
    elif dimension == "vehicle":
        group_cols = [Vehicle.vehicle_id, Vehicle.registration_number]
        label = lambda r: r.registration_number or r.vehicle_id  # noqa: E731
    else:
        group_cols = [Vehicle.transporter]
        label = lambda r: r.transporter  # noqa: E731

    on_time = case((Trip.actual_arrival <= Trip.planned_arrival, 1), else_=0)
    query = (
        select(*group_cols,
               func.count(Trip.id).label("trips"),
               func.sum(on_time).label("on_time"),
               func.avg(Trip.actual_duration_hours - Trip.planned_duration_hours).label("avg_delay"),
               func.sum(Trip.distance_km).label("distance"),
               func.sum(Trip.fuel_liters).label("fuel"))
        .where(Trip.trip_status == COMPLETED)
        .group_by(*group_cols)
        .having(func.count(Trip.id) >= 10)
    )
    if joins_vehicle:
        query = query.join(Vehicle, Vehicle.vehicle_id == Trip.vehicle_id)
    rows = db.session.execute(query.order_by(func.count(Trip.id).desc()).limit(limit)).all()

    out = []
    for r in rows:
        m = r._mapping
        fuel = m["fuel"] or 0
        out.append({
            "label": label(r),
            "trips": m["trips"],
            "otd_rate": round(m["on_time"] / m["trips"] * 100, 1),
            "avg_delay_hours": round(m["avg_delay"] or 0, 1),
            "distance_km": round(m["distance"] or 0, 0),
            "fuel_efficiency_km_l": round(m["distance"] / fuel, 2) if fuel else None,
        })
    return out


def delay_distribution() -> dict:
    """Histogram of trip delays (actual - planned) in hours."""
    delays = db.session.execute(
        select(Trip.actual_duration_hours - Trip.planned_duration_hours)
        .where(Trip.trip_status == COMPLETED)
    ).scalars().all()
    buckets = {"< -2h": 0, "-2h..0": 0, "0..4h": 0, "4h..12h": 0, "12h..24h": 0, "> 24h": 0}
    for d in delays:
        if d is None:
            continue
        if d < -2:
            buckets["< -2h"] += 1
        elif d < 0:
            buckets["-2h..0"] += 1
        elif d < 4:
            buckets["0..4h"] += 1
        elif d < 12:
            buckets["4h..12h"] += 1
        elif d < 24:
            buckets["12h..24h"] += 1
        else:
            buckets["> 24h"] += 1
    return {"labels": list(buckets.keys()), "counts": list(buckets.values())}


def planned_vs_actual_by_month(months: int = 12) -> dict:
    end = datetime.now()
    start = end - timedelta(days=months * 31)
    month = func.strftime("%Y-%m", Trip.departure_time).label("month")
    rows = db.session.execute(
        select(month,
               func.avg(Trip.planned_duration_hours).label("planned"),
               func.avg(Trip.actual_duration_hours).label("actual"))
        .where(Trip.trip_status == COMPLETED, Trip.departure_time >= start)
        .group_by(month).order_by(month)
    ).all()
    return {
        "labels": [r.month for r in rows],
        "planned": [round(r.planned, 1) for r in rows],
        "actual": [round(r.actual, 1) for r in rows],
    }

File: app/static/js/transport.js
/**
 * LogiSight — Transport performance page
 */

const PLOT = (el, data, layout = {}) =>
    Plotly.newPlot(prepPlotEl(el), data, {
        margin: { l: 48, r: 12, t: 8, b: 40 },
        font: { family: "Inter, sans-serif", size: 11, color: "#6b7a90" },
        paper_bgcolor: "rgba(0,0,0,0)", plot_bgcolor: "rgba(0,0,0,0)",
        ...layout,
    }, { displayModeBar: false, responsive: true });

const TRIP_STATUS = {
    PLANNED: ["Planifiée", "neutral"], IN_PROGRESS: ["En cours", "info"],
    COMPLETED: ["Terminée", "success"], CANCELLED: ["Annulée", "neutral"],
};

let tripPage = 1;

document.addEventListener("DOMContentLoaded", () => {
    GlobalFilters.onChange(refresh);
    refresh();
});

async function refresh() {
    const params = GlobalFilters.apiParams(["origin", "destination", "customer_id", "cargo_type", "status"]);
    loadSummary(params);
    loadCharts();
    loadTrips();
}

async function loadSummary(params) {
    try {
        const res = await API.get("/api/v1/transport/summary", params);
        const d = res.data;
        const kpis = [
            { label: "Tournées totales", value: fmt.num(d.total_trips), icon: "🚚", color: "blue" },
            { label: "Tournées à l'heure", value: fmt.num(d.on_time_trips), icon: "✅", color: "green" },
            { label: "Ponctualité", value: fmt.pct(d.on_time_rate), icon: "🕒", color: "cyan" },
            { label: "Retard moyen", value: d.avg_delay_hours != null ? fmt.hours(d.avg_delay_hours) : "—", icon: "⚠️", color: "orange" },
            { label: "Vitesse moyenne", value: d.avg_speed_kmh ? fmt.num(d.avg_speed_kmh, 1) + " km/h" : "—", icon: "📏", color: "blue" },
            { label: "Distance totale", value: fmt.km(d.total_distance_km), icon: "🛣️", color: "cyan" },
            { label: "Carburant total", value: fmt.num(d.total_fuel_liters, 0) + " L", icon: "⛽", color: "orange" },
            { label: "Rendement carburant", value: d.fuel_efficiency_km_l ? fmt.num(d.fuel_efficiency_km_l, 2) + " km/L" : "—", icon: "🌱", color: "green" },
        ];
        document.getElementById("transport-kpis").innerHTML = kpis.map(renderKpiCard).join("");
    } catch (e) {
        document.getElementById("transport-kpis").innerHTML = `<div class="card">${renderError(e.message)}</div>`;
    }
}

async function loadCharts() {
    try {
        const [pva, delays, routes, vehicles, transporters] = await Promise.all([
            API.get("/api/v1/transport/planned-vs-actual"),
            API.get("/api/v1/transport/delays"),
            API.get("/api/v1/transport/routes"),
            API.get("/api/v1/transport/vehicles"),
            API.get("/api/v1/transport/transporters"),
        ]);

        PLOT("chart-pva", [
            { x: pva.data.labels, y: pva.data.planned, type: "scatter", mode: "lines+markers", name: "Planifié (h)", line: { color: CHART_COLORS[0], width: 2 } },
            { x: pva.data.labels, y: pva.data.actual, type: "scatter", mode: "lines+markers", name: "Réel (h)", line: { color: CHART_COLORS[3], width: 2 } },
        ], { legend: { orientation: "h", y: -0.25 } });

        PLOT("chart-delays", [{
            x: delays.data.labels, y: delays.data.counts, type: "bar",
            marker: { color: [CHART_COLORS[1], CHART_COLORS[1], CHART_COLORS[0], CHART_COLORS[2], CHART_COLORS[2], CHART_COLORS[3]] },
        }]);

        const dimChart = (id, rows) => {
            const top = rows.slice(0, 6);
            PLOT(id, [{
                x: top.map(r => r.label), y: top.map(r => r.otd_rate), type: "bar",
                name: "Ponctualité (%)",
                marker: { color: top.map(r => r.otd_rate >= 75 ? CHART_COLORS[1] : r.otd_rate >= 60 ? CHART_COLORS[2] : CHART_COLORS[3]) },
                hovertemplate: "%{x}<br>Ponctualité: %{y}%<extra></extra>",
            }], { xaxis: { tickangle: -25, automargin: true } });
        };
        dimChart("chart-routes", routes.data);
        dimChart("chart-vehicles", vehicles.data);
        dimChart("chart-transporters", transporters.data);
    } catch (e) {
        ["chart-pva", "chart-delays", "chart-routes", "chart-vehicles", "chart-transporters"]
            .forEach(id => document.getElementById(id).innerHTML = renderError(e.message));
    }
}

async function loadTrips() {
    const el = document.getElementById("trips-table");
    try {
        const params = {
            start_date: document.getElementById("f-start").value,
            end_date: document.getElementById("f-end").value,
            page: tripPage, per_page: 10,
        };
        const res = await API.get("/api/v1/transport/trips", params);
        const rows = res.data;
        if (!rows.length) { el.innerHTML = renderEmpty("Aucune tournée pour ces filtres."); return; }
        const cols = [
            { key: "trip_id", label: "Tournée" },
            { key: "shipment_id", label: "Expédition", render: r => `<a href="/shipments/${r.shipment_id}">${r.shipment_id}</a>` },
            { label: "Véhicule", render: r => r.vehicle_registration || r.vehicle_id },
            { label: "Transporteur", render: r => r.transporter || "—" },
            { label: "Trajet", render: r => `${r.origin} → ${r.destination}` },
            { label: "Départ", render: r => fmt.date(r.departure_time) },
            { label: "Prévu", num: true, render: r => fmt.hours(r.planned_duration_hours) },
            { label: "Réel", num: true, render: r => r.actual_duration_hours ? fmt.hours(r.actual_duration_hours) : "—" },
            { label: "Distance", num: true, render: r => fmt.km(r.distance_km) },
            { label: "Statut", render: r => badge(...(TRIP_STATUS[r.trip_status] || [r.trip_status, "neutral"])) },
        ];
        el.innerHTML = renderTable(cols, rows);
        document.getElementById("trips-pagination").innerHTML = renderPagination(res.meta);
        document.querySelectorAll("#trips-pagination button[data-page]").forEach(b =>
            b.addEventListener("click", () => { tripPage = +b.dataset.page; loadTrips(); }));
    } catch (e) {
        el.innerHTML = renderError(e.message);
    }
}
